src/combinatorics/composition.lean

```diff
@@ -0,0 +1,676 @@
+/-
+Copyright (c) 2020 Sébastien Gouëzel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel
+-/
+
+import data.fintype.card tactic.omega
+
+/-!
+# Compositions
+
+A composition of an integer `n` is a decomposition `n = i₀ + ... + i_{k-1}` of `n` into a sum of
+positive integers. Combinatorially, it corresponds to a decomposition of `{0, ..., n-1}` into
+non-empty blocks of consecutive integers, where the `iⱼ` are the lengths of the blocks.
+This notion is closely related to that of a partition of `n`, but in a composition of `n` the
+order of the `iⱼ`s matters.
+
+We implement two different structures covering these two viewpoints on compositions. The first
+one, made of a list of positive integers summing to `n`, is the main one and is called
+`composition n`. The second one is useful for combinatorial arguments (for instance to show that
+the number of compositions of `n` is `2^(n-1)`). It is given by a subset of `{0, ..., n}`
+containing `0` and `n`, where the elements of the subset (but `n`) correspond to the leftmost
+points of each block. The main API is built on `composition n`, and we provide an equivalence
+between the two types.
+
+## Main functions
+
+* `c : composition n` is a structure, made of a list of positive integers summing to `n`.
+* `composition_card` states that the cardinality of `composition n` is exactly
+  `2^(n-1)`, which is proved by constructing an equiv with `composition_as_set n` (see below), which
+  is itself in bijection with the subsets of `fin (n-1)` (this holds even for `n = 0`, where `-` is
+  nat subtraction).
+
+Let `c : composition n` be a composition of `n`. Then
+* `c.blocks_pnat` is the list of blocks in `c`, as positive integers
+* `c.blocks` is the list of blocks in `c`, as elements of `ℕ`
+* `c.length` is the number of blocks in the composition;
+* `c.blocks_fun : fin c.length → ℕ` is the realization of `c.blocks` as a function on
+  `fin c.length`. This is the main object when using compositions to understand the composition of
+    analytic functions.
+* `c.size_up_to : ℕ → ℕ` is the sum of the size of the blocks up to `i`.;
+* `c.embedding i : fin (c.blocks_fun i) → fin n` is the increasing embedding of the `i`-th block in
+  `fin n`;
+* `c.index j`, for `j : fin n`, is the index of the block containing `j`.
+
+We turn two to the second viewpoint on compositions, that we realize as a finset of `fin (n+1)`.
```

Comment:
Not sure what "two" is doing in this sentence.

Reply:
Nothing, really (too much refactoring of the docs...)